lib/AtomicCache.php

```diff
@@ -0,0 +1,283 @@
+     *
+     * @param string   $key   Cache key.
+     * @param mixed    $value Value to cache.
+     * @param int|null $ttl   Timeout in seconds. The default `null` $ttl value indicates no timeout.
+     *
+     * @return Promise<mixed> Resolves either with the current value or with the newly set value. Fails with a
+     * CacheException or SerializationException on failure.
+     *
+     * @throws CacheException
+     * @throws SerializationException
+     */
+    public function setIfAbsent(string $key, $value, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $value, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                $currentValue = yield $this->cache->get($key);
+
+                if ($currentValue !== null) {
+                    return $currentValue;
+                }
+
+                yield $this->cache->set($key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+
+            return $value;
+        });
+    }
+
+    private function lock(string $key): \Generator
+    {
+        try {
+            return yield $this->mutex->acquire($key);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown when obtaining the lock for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+    }
+
+    private function create(callable $create, string $key, $value, ?int $ttl): \Generator
+    {
+        try {
+            $value = yield call($create, $key, $value);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown while creating the value for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+
+        yield $this->cache->set($key, $value, $ttl);
+
+        return $value;
+    }
+
+    /**
+     * Returns the cached value for the key or the given default value if the key does not exist.
+     *
+     * @param string     $key     Cache key.
+     * @param mixed|null $default Default value returned if the key does not exist. Null by default.
+     *
+     * @return Promise<mixed|null> Resolved with null iff $default is null.
+     *
+     * @throws CacheException
+     * @throws SerializationException
+     *
+     * @see SerializedCache::get()
+     */
+    public function get(string $key, $default = null): Promise
+    {
+        return call(function () use ($key, $default): \Generator {
+            $value = yield $this->cache->get($key);
+
+            if ($value === null) {
+                return $default;
+            }
+
+            return $value;
+        });
+    }
+
+    /**
+     * The lock is obtained for the key before deleting the key.
+     *
+     * @param string $key
+     *
+     * @return Promise<bool>
```

Comment:
The type should be `Promise<bool|null>` here and in `Cache` according to the documentation comment.

Reply:
True but honestly I hate that because then I have to check the return type for every delete call. I'd much rather if the method threw an exception instead.

Reply:
@enumag You only need to check it if you need the information whether the entry really existed before deleting. Throwing an exception if this information can't be provided is inappropriate, as it's not an error condition. Some implementations might simply not be able to provide such information on delete calls.

Reply:
@kelunik Updated here and in `Cache` and `SerializedCache`.